python/ray/data/impl/pandas_block.py

```diff
@@ -0,0 +1,189 @@
+
+
+class PandasBlockBuilder(TableBlockBuilder[T]):
+    def __init__(self):
+        if pandas is None:
+            raise ImportError("Run `pip install pandas` for Pandas support")
+        TableBlockBuilder.__init__(self, pandas.DataFrame)
+
+    def _table_from_pydict(self, columns: Dict[str, List[Any]]) -> Block:
+        return pandas.DataFrame(columns)
+
+    def _concat_tables(self, tables: List[Block]) -> Block:
+        return pandas.concat(tables, ignore_index=True)
+
+    @staticmethod
+    def _empty_table() -> "pandas.DataFrame":
+        return pandas.DataFrame()
+
+
+# This is to be compatible with pyarrow.lib.schema
+# TODO (kfstorm): We need a format-independent way to represent schema.
+PandasBlockSchema = collections.namedtuple("PandasBlockSchema",
+                                           ["names", "types"])
+
+
+class PandasBlockAccessor(TableBlockAccessor):
+    def __init__(self, table: "pandas.DataFrame"):
+        if pandas is None:
+            raise ImportError("Run `pip install pandas` for Pandas support")
+        TableBlockAccessor.__init__(self, table)
+
+    def _create_table_row(self, row: "pandas.DataFrame") -> PandasRow:
+        return PandasRow(row)
+
+    def slice(self, start: int, end: int, copy: bool) -> "pandas.DataFrame":
+        view = self._table[start:end]
+        if copy:
+            view = view.copy(deep=True)
+        return view
+
+    def random_shuffle(self, random_seed: Optional[int]) -> List[T]:
+        return self._table.sample(frac=1, random_state=random_seed)
+
+    def schema(self) -> Any:
+        dtypes = self._table.dtypes
+        schema = PandasBlockSchema(
+            names=dtypes.index.tolist(), types=dtypes.values.tolist())
+        # Column names with non-str types of a pandas DataFrame is not
+        # supported by Ray Dataset.
+        if any(not isinstance(name, str) for name in schema.names):
+            raise ValueError(
+                "A Pandas DataFrame with column names of non-str types"
+                " is not supported by Ray Dataset. Column names of this"
+                f" DataFrame: {schema.names!r}.")
+        return schema
+
+    def to_pandas(self) -> "pandas.DataFrame":
+        return self._table
+
+    def to_numpy(self, column: str = None) -> np.ndarray:
+        if not column:
+            raise ValueError(
+                "`column` must be specified when calling .to_numpy() "
+                "on Pandas blocks.")
+        if column not in self._table.columns:
+            raise ValueError(
+                "Cannot find column {}, available columns: {}".format(
+                    column, self._table.columns.tolist()))
+        return self._table[column].to_numpy()
+
+    def to_arrow(self) -> "pyarrow.Table":
+        import pyarrow
+        return pyarrow.table(self._table)
+
+    def num_rows(self) -> int:
+        return self._table.shape[0]
+
+    def size_bytes(self) -> int:
+        # TODO (kfstorm): Should we count index?
```

Comment:
I think it's good to (since this is used for size estimation for splitting, etc.)

Reply:
+1